platform/cervantes/koreader.sh

```diff
@@ -91,20 +94,101 @@ ENTER_QBOOKAPP=87
+        ko_update_check
+    fi
 
     # run KOReader
     ./reader.lua "$@" >>crash.log 2>&1
     RETURN_VALUE=$?
 
+    if [ ${RETURN_VALUE} -ne 0 ] && [ "${RETURN_VALUE}" -ne "${ENTER_USBMS}" ] && [ "${RETURN_VALUE}" -ne "${ENTER_QBOOKAPP}" ] && [ "${RETURN_VALUE}" -ne "${RESTART_KOREADER}" ]; then
+        # Increment the crash counter
+        CRASH_COUNT=$((CRASH_COUNT + 1))
+        CRASH_TS=$(date +'%s')
+        # Reset it to a first crash if it's been a while since our last crash...
+        if [ $((CRASH_TS - CRASH_PREV_TS)) -ge 20 ]; then
+            CRASH_COUNT=1
+        fi
+
+        # Check if the user requested to always abort on crash
+        if grep -q '\["dev_abort_on_crash"\] = true' 'settings.reader.lua' 2>/dev/null; then
+            ALWAYS_ABORT="true"
+            # In which case, make sure we pause on *every* crash
+            CRASH_COUNT=1
+        else
+            ALWAYS_ABORT="false"
+        fi
+
+        # Show a fancy bomb on screen
+        viewWidth=600
+        viewHeight=800
+        FONTH=16
+        eval "$(./fbink -e | tr ';' '\n' | grep -e viewWidth -e viewHeight -e FONTH | tr '\n' ';')"
+        # Compute margins & sizes relative to the screen's resolution, so we end up with a similar layout, no matter the device.
+        # Height @ ~56.7%, w/ a margin worth 1.5 lines
+        bombHeight=$((viewHeight / 2 + viewHeight / 15))
+        bombMargin=$((FONTH + FONTH / 2))
+        # With a little notice at the top of the screen, on a big gray screen of death ;).
+        ./fbink -q -b -c -B GRAY9 -m -y 1 "Don't Panic! (Crash n°${CRASH_COUNT} -> ${RETURN_VALUE})"
+        if [ ${CRASH_COUNT} -eq 1 ]; then
+            # Warn that we're waiting on a tap to continue...
+            ./fbink -q -b -O -m -y 2 "Tap the screen to continue."
+        fi
+        # U+1F4A3, the hard way, because we can't use \u or \U escape sequences...
+        ./fbink -q -b -O -m -t regular=./fonts/freefont/FreeSerif.ttf,px=${bombHeight},top=${bombMargin} -- $'\xf0\x9f\x92\xa3'
+        # And then print the tail end of the log on the bottom of the screen...
+        crashLog="$(tail -n 25 crash.log | sed -e 's/\t/    /g')"
+        # The idea for the margins being to leave enough room for an fbink -Z bar, small horizontal margins, and a font size based on what 6pt looked like @ 265dpi
+        ./fbink -q -b -O -t regular=./fonts/droid/DroidSansMono.ttf,top=$((viewHeight / 2 + FONTH * 2 + FONTH / 2)),left=$((viewWidth / 60)),right=$((viewWidth / 60)),px=$((viewHeight / 64)) -- "${crashLog}"
+        # So far, we hadn't triggered an actual screen refresh, do that now, to make sure everything is bundled in a single flashing refresh.
+        ./fbink -q -f -s
+        # Cue a lemming's faceplant sound effect!
+
+        {
+            echo "!!!!"
+            echo "Uh oh, something went awry... (Crash n°${CRASH_COUNT}: $(date +'%x @ %X'))"
+            echo "Running FW $(xargs </.version) on Linux $(uname -r) ($(uname -v))"
```

Comment:
Does anyone know a better way to get the FW version on Cervantes? On my cervantes 4 this returns "246 71d478", which may or may not be useful.

Reply:
I don't know anything about the platform as such but 246 looks like it should be simple and meaningful?

Reply:
There's no way to retrieve the FW version. You can:

1. print the HWConfig version: `echo "Running FW $(ntxinfo /dev/mmcblk0 | grep version_name | cut -d ":" -f2)  on Linux $(uname -r) ($(uname -v))"`

2. print the debian version: `echo "Running Debian $(cat /etc/debian_version). Linux $(uname -r) ($(uname -v))"`

Reply:
And which of these three options are actually useful to show? We could show any combination of them or not show anything at all.

Reply:
FWIW, this made sense on Kobo, it might not here, so I wouldn't necessarily be overly attached to it. I'll let @pazos be the judge of that ;).

Reply:
> And which of these three options are actually useful to show?

The company is dead and each supported device has **one** FW version. So

> not show anything at all.